lms/djangoapps/courseware/models.py

```diff
@@ -169,25 +166,71 @@ class Meta(object):
     grade = models.FloatField(null=True, blank=True)
     max_grade = models.FloatField(null=True, blank=True)
 
-    @receiver(post_save, sender=StudentModule)
-    def save_history(sender, instance, **kwargs):  # pylint: disable=no-self-argument, unused-argument
+    @property
+    def csm(self):
         """
-        Checks the instance's module_type, and creates & saves a
-        StudentModuleHistory entry if the module_type is one that
-        we save.
+        Finds the StudentModule object for this history record, even if our data is split
+        across multiple data stores.  Django does not handle this correctly with the built-in
+        student_module property.
         """
-        if instance.module_type in StudentModuleHistory.HISTORY_SAVING_TYPES:
-            history_entry = StudentModuleHistory(student_module=instance,
-                                                 version=None,
-                                                 created=instance.modified,
-                                                 state=instance.state,
-                                                 grade=instance.grade,
-                                                 max_grade=instance.max_grade)
-            history_entry.save()
+        return StudentModule.objects.get(pk=self.student_module_id)
+
+    @staticmethod
+    def get_history(student_modules):
+        """
+        Find history objects across multiple backend stores for a given StudentModule
+        """
+
+        history_entries = []
+        # If we turn off reading from multiple history tables, then we don't want to read from
+        # StudentModuleHistory anymore, we believe that all history is in the Extended table.
+        if not settings.FEATURES.get('ENABLE_READING_FROM_MULTIPLE_HISTORY_TABLES'):
+            # we want to save later SQL queries on the model which allows us to prefetch
+            history_entries += StudentModuleHistory.objects.prefetch_related('student_module').filter(
+                student_module__in=student_modules
+            ).order_by('-id')
+
+        if settings.FEATURES.get('ENABLE_CSMH_EXTENDED'):
+            history_entries += coursewarehistoryextended.models.StudentModuleHistoryExtended.objects.filter(
+                # Django will sometimes try to join to courseware_studentmodule
+                # so just do an in query
+                student_module__in=[module.id for module in student_modules]
+            ).order_by('-id')
```

Comment:
Your question made me realize there was a logic error here which I didn't catch.
I've pushed a fix for that as well as unit tests to test the 4 configurations of these two feature flags.
I can add a startup warning, but a) I hope that the second flag will eventually be deprecated and removed b) this is not a data loss configuration, it is a data display failure and fixing your flags will cause it to be returned again.
If we decide to go with a warning, I suspect it will result in a check_for_misconfigurations block in startup.py, although I'm not sure if that catches all places.